rfd/0xxx-partial-cache-healthiness.md

```diff
@@ -0,0 +1,322 @@
+2. A totally new resource gets added to cache in a new teleport version. Requests forwarded to Auth API
+   would fail with `NotImplemented`, but the remaining cache would not be affected. The component responsible
+   for the failed request would have a chance to handle the error and gracefully degrade its functionality.
+
+Relevant issue: [#21586](https://github.com/gravitational/teleport/issues/21586)
+
+
+## Details
+
+### How cache currently works
+
+`Cache` interface shares numerous methods with Auth API Client interface `ClientI` with the intention that an instance
+of Cache could be used transparently instead of the client. Cache has a reference to the actual API client that
+it can be used as a source of truth.
+
+Internally, the implementation of Cache includes implementations of all relevant upstream services
+running off of an in-memory storage backend. It makes use of the `Events` interface implemented by the API client
+to subscribe to *relevant* changes happening on the actual Auth service backend, and replay them into the memory backend.
+Each downstream service has a specific set of resource kinds that are relevant to it and events for which it wants
+to follow.
+
+```go
+type Events interface {
+    NewWatcher(ctx context.Context, watch Watch) (Watcher, error)
+}
+```
+
+For synchronization purposes, a special `OpInit` event with no payload always comes first on the stream.
+
+Upon receipt of `OpInit`, cache fetches all relevant resources using the API client, puts them into the in-memory storage
+and starts replaying the event stream. Only at this point cache transitions into a healthy state and can be used
+by the readers. While being unhealthy, cache routes all read requests to the original API Client. It can become unhealthy
+again if the event stream gets interrupted.
+
+Currently, if some of the requested kinds are unsupported by Auth API, cache never becomes healthy.
+The goal of this RFD is to propose a way to avoid that.
+
+Auth API client is not the only thing that implements the `Events` interface and can produce a stream of events.
+It's implemented by multiple types, including `local.EventsService` which generates events from the storage backend and
+`Cache` itself which helps to fan out an event stream to multiple consumers. There are also a few implementations
+of `Events` that wrap each other to provide validation.
+
+
+### General idea
+
+When requesting a stream of events from Auth service, the cache should set a new flag on the request to opt in for the new partial
+success mode. In this mode, the request won't fail if some of the requested kinds cannot be watched. Instead, it will succeed and the first
+`OpInit` event on the resulting stream will contain the list of kinds and subkinds rejected by the watch operation. The list will
+be attached to a new special resource type associated with the event. For backwards compatibility, the watch request
+should fail if the new flag wasn't set on it.
```

Comment:
Actually, even before that, if an old provider doesn't recognize some of the requested kinds, it will return a failure, but if it recognizes all of them, it doesn't matter if it's old or new.